src/scratchOrgFeatureDeprecation.ts

```diff
@@ -0,0 +1,123 @@
+  simpleFeatureMapping: {
+    SALESWAVE: ['DEVELOPMENTWAVE'],
+    SERVICEWAVE: ['DEVELOPMENTWAVE'],
+  },
+  quantifiedFeatureMapping: {},
+  deprecatedFeatures: [
+    'EXPANDEDSOURCETRACKING',
+    'LISTCUSTOMSETTINGCREATION',
+    'AppNavCapabilities',
+    'EditInSubtab',
+    'OldNewRecordFlowConsole',
+    'OldNewRecordFlowStd',
+    'DesktopLayoutStandardOff',
+    'SplitViewOnStandardOff',
+    'PopOutUtilities',
+  ],
+};
+
+interface FeatureTypes {
+  simpleFeatureMapping: { [key: string]: string[] };
+  quantifiedFeatureMapping: Record<string, string>;
+  deprecatedFeatures: string[];
+}
+
+export class ScratchOrgFeatureDeprecation {
+  private featureTypes: FeatureTypes;
+  // Allow override for testing.
+  public constructor(options: FeatureTypes = FEATURE_TYPES) {
+    this.featureTypes = options;
+    this.featureTypes.deprecatedFeatures = this.featureTypes.deprecatedFeatures.map((deprecatedFeature) =>
+      deprecatedFeature.toUpperCase()
+    );
+    // Make all of the keys in simpleFeatureMapping upper case.
+    const sfm: { [key: string]: string[] } = {};
+    Object.keys(this.featureTypes.simpleFeatureMapping).forEach((key) => {
+      sfm[key.toUpperCase()] = this.featureTypes.simpleFeatureMapping[key];
+    });
+    this.featureTypes.simpleFeatureMapping = sfm;
+  }
+
+  /**
+   * Gets list of feature warnings that should be logged
+   *
+   * @param features The requested features.
+   * @returns List of string feature warnings.
+   */
+  public getFeatureWarnings(features: string | string[]): string[] {
+    /* Get warning messages for deprecated features and feature mappings.*/
+    const featureWarningMessages: string[] = [];
+    if (features) {
+      const requestedFeatures = (isString(features) ? features : features.join(';')).toUpperCase();
+
+      /* If a public quantified feature is defined without a quantity, throw a warning.*/
+      Object.keys(this.featureTypes.quantifiedFeatureMapping).forEach((key) => {
+        if (new RegExp(`${key};|${key},|${key}$`, 'i').test(requestedFeatures)) {
+          featureWarningMessages.push(
+            messages.getMessage('quantifiedFeatureWithoutQuantityWarning', [
+              key,
+              this.featureTypes.quantifiedFeatureMapping[key],
+            ])
+          );
+        }
+      });
+      /* If a simply mapped feature is defined, log a warning.*/
+      Object.keys(this.featureTypes.simpleFeatureMapping).forEach((key) => {
+        if (new RegExp(`${key};|${key},|${key}$`, 'i').test(requestedFeatures)) {
+          const tokens = '[' + this.featureTypes.simpleFeatureMapping[key].map((v) => "'" + v + "'").join(',') + ']';
+          featureWarningMessages.push(messages.getMessage('mappedFeatureWarning', [key, tokens]));
+        }
+      });
+      /* If a deprecated feature is identified as deprecated, throw a warning.*/
+      this.featureTypes.deprecatedFeatures.forEach((deprecatedFeature) => {
```

Comment:
that looks like a `filter`/`map`
`featureWarningMessages.push(
    ...this.featureTypes.deprecatedFeatures
      .filter(deprecated => requestedFeatures.includes(deprecated))
      .map(deprecated => messages.getMessage('deprecatedFeatureWarning', [deprecatedFeature]))
)`